src/compiler/marks.ts

```diff
@@ -1,111 +1,84 @@
 import {Model} from './Model';
 import {COLUMN, ROW, X, Y, COLOR, TEXT, SIZE, SHAPE} from '../channel';
+import {AREA, BAR, LINE, POINT, TEXT as TEXTMARKS, TICK, CIRCLE, SQUARE} from '../marktype';
 import {QUANTITATIVE} from '../type';
 
 // https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#11-ambient-declarations
 declare var exports;
 
+/* mapping from vega-lite's mark types to vega's mark types */
+const MARKTYPES_MAP = {
+  bar: 'rect',
+  tick: 'rect',
+  point: 'symbol',
+  line: 'line',
+  area: 'area',
+  text: 'text',
+  circle: 'symbol',
+  square: 'symbol'
+};
+
 export function compileMarks(model: Model, layout, style) {
-  var defs = [],
-    mark = exports[model.marktype()],
-    from = model.dataTable();
-
-  // to add a background to text, we need to add it before the text
-  if (model.marktype() === TEXT && model.has(COLOR)) {
-    var bg = {
-      x: {value: 0},
-      y: {value: 0},
-      x2: {value: layout.cellWidth},
-      y2: {value: layout.cellHeight},
-      fill: {scale: COLOR, field: model.fieldRef(COLOR)}
-    };
-    defs.push({
-      type: 'rect',
-      from: {data: from},
-      properties: {enter: bg, update: bg}
-    });
-  }
+  const marktype = model.marktype();
+  const from = {from: model.dataTable()};
+  let defs = [];
+
+  switch (marktype) {
+    case TEXTMARKS:
+      if (model.has(COLOR)) {
+        defs.push({
+          type: 'rect',
+          from: from,
+          properties: {update: properties.textBackground(model, layout)}
+        });
+      }
+      break;
+    case LINE:
+    case AREA:
+      // TODO: move subfacets and line sort logic here
+  };
 
-  // add the mark def for the main thing
-  var p = mark.prop(model, layout, style);
   defs.push({
-    type: mark.type,
-    from: {data: from},
-    properties: {update: p}
+    type: MARKTYPES_MAP[marktype],
+    from: from,
+    properties: {
+      update: properties[marktype](model, layout, style)
+    }
   });
 
   return defs;
 }
 
-export const bar = {
-  type: 'rect',
-  prop: bar_props
-};
-
-export const line = {
-  type: 'line',
-  prop: line_props
-};
-
-export const area = {
-  type: 'area',
-  prop: area_props
-};
-
-export const tick = {
-  type: 'rect',
-  prop: tick_props
-};
-
-export const circle = {
-  type: 'symbol',
-  prop: filled_point_props('circle')
-};
-
-export const square = {
-  type: 'symbol',
-  prop: filled_point_props('square')
-};
-
-export const point = {
-  type: 'symbol',
-  prop: point_props
-};
-
-export const text = {
-  type: 'text',
-  prop: text_props
-};
-
-function bar_props(e: Model, layout, style) {
+export namespace properties {
```

Comment:
@domoritz  I start wondering if we should organize this in the opposite way e.g.,
```ts
export namespace bar  {
   export function properties() {
   }
}
```

This way, we can easily add `label` to each mark's namespace.